### PR TITLE
feat: implement validation webhook for Queue

### PR DESCRIPTION
**Pull request by knight42**
Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Ref #171

### CODE REVIEW

File: apis/kueue/v1alpha1/queue_webhook_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	. "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	testingutil "sigs.k8s.io/kueue/pkg/util/testing"
+)
+
+// Rename the package to avoid circular dependencies which is caused by "sigs.k8s.io/kueue/pkg/util/testing".
+// See also: https://github.com/golang/go/wiki/CodeReviewComments#import-dot
+
+func TestValidateQueueUpdate(t *testing.T) {
+	const (
+		objName = "name"
```
Comment (kerthcet) [inline apis/kueue/v1alpha1/queue_webhook_test.go:35]: replace with a more meaningful name.

File: apis/kueue/v1alpha1/queue_webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+			},
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			errList := ValidateQueueUpdate(tc.before, tc.after)
+			if len(errList) == 0 {
+				t.Fatalf("Unexpected error: %v, want %v", errList, tc.wantErr)
+			}
+			if diff := cmp.Diff(tc.wantErr[0], errList[0], cmpopts.IgnoreFields(field.Error{}, "Detail", "BadValue")); diff != "" {
```
Comment (kerthcet) [inline apis/kueue/v1alpha1/queue_webhook_test.go:56]: As suggested before.

File: test/integration/webhook/v1alpha1/queue_test.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/kueue/pkg/util/testing"
+)
+
+const queueName = "queue-test"
+
+var _ = ginkgo.Describe("Queue validating webhook", func() {
+	ginkgo.Context("When updating a Queue", func() {
```
Comment (kerthcet) [inline test/integration/webhook/v1alpha1/queue_test.go:27]: Add a normal path as you do in workload validation?

File: apis/kueue/v1alpha1/queue_webhook.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/validate-kueue-x-k8s-io-v1alpha1-queue,mutating=false,failurePolicy=fail,sideEffects=None,groups=kueue.x-k8s.io,resources=queues,verbs=create;update,versions=v1alpha1,name=vqueue.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &Queue{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Queue) ValidateCreate() error {
+	return nil
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/queue_webhook.go:43]: We can validate that the field is a valid object name `validation.NameIsDNSSubdomain` from https://github.com/kubernetes/kubernetes/blob/8c8a4cf3e4a18e97359ce750530a4fa27bbd3b88/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#L60

File: apis/kueue/v1alpha1/queue_webhook.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *Queue) ValidateUpdate(old runtime.Object) error {
+	queuelog.V(5).Info("validate update", "queue", klog.KObj(r))
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/queue_webhook.go:48]:
```suggestion
	queuelog.V(5).Info("Validating update", "queue", klog.KObj(r))
```

File: apis/kueue/v1alpha1/queue_webhook.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var queuelog = ctrl.Log.WithName("queue-webhook")
+
+func (r *Queue) SetupWebhookWithManager(mgr ctrl.Manager) error {
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/queue_webhook.go:31]:
```suggestion
func (q *Queue) SetupWebhookWithManager(mgr ctrl.Manager) error {
```

File: apis/kueue/v1alpha1/queue_webhook_test.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+func TestValidateQueueUpdate(t *testing.T) {
+	const (
+		objName = "name"
+		objNs   = "ns"
+	)
+	testCases := map[string]struct {
+		before, after *Queue
+		wantErr       field.ErrorList
+	}{
+		"clusterQueue cannot be updated": {
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/queue_webhook_test.go:42]: add a test case for changes in other fields (like status).

File: apis/kueue/v1alpha1/queue_webhook_test.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	. "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	testingutil "sigs.k8s.io/kueue/pkg/util/testing"
+)
+
+// Rename the package to avoid circular dependencies which is caused by "sigs.k8s.io/kueue/pkg/util/testing".
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/queue_webhook_test.go:30]: Put this above the package name (but leave an empty line in-between).

File: apis/kueue/v1alpha1/queue_webhook_test.go (hunk 2 above)
Comment (alculquicondor) [inline apis/kueue/v1alpha1/queue_webhook_test.go:56]:
```suggestion
			if diff := cmp.Diff(tc.wantErr, errList, cmpopts.IgnoreFields(field.Error{}, "Detail", "BadValue")); diff != "" {
```

File: apis/kueue/v1alpha1/queue_webhook.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	return ValidateQueueUpdate(q, old.(*Queue)).ToAggregate()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (q *Queue) ValidateDelete() error {
+	return nil
+}
+
+func ValidateQueueCreate(q *Queue) field.ErrorList {
+	return apivalidation.ValidateObjectMetaAccessor(q.GetObjectMeta(), true, apivalidation.NameIsDNSLabel, field.NewPath("metadata"))
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/kueue/v1alpha1/queue_webhook.go:59]: And the method to use to validate is `NameIsDNSSubdomain` https://github.com/kubernetes/kubernetes/blob/8c8a4cf3e4a18e97359ce750530a4fa27bbd3b88/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#L60

File: apis/kueue/v1alpha1/queue_webhook_test.go (hunk 7 above)
Comment (knight42) [reply apis/kueue/v1alpha1/queue_webhook_test.go:68]: This should be fixed now.

File: apis/kueue/v1alpha1/queue_webhook.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (q *Queue) ValidateDelete() error {
+	return nil
+}
+
+func ValidateQueueCreate(q *Queue) field.ErrorList {
+	var allErrs field.ErrorList
+	for _, msg := range apivalidation.NameIsDNSSubdomain(string(q.Spec.ClusterQueue), false) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "clusterQueue"), q.Spec.ClusterQueue, msg))
```
Comment (ahg-g) [inline apis/kueue/v1alpha1/queue_webhook.go:61]: create the path once before the for loop

File: test/integration/webhook/v1alpha1/queue_test.go (hunk 3 above)
Comment (ahg-g) [reply test/integration/webhook/v1alpha1/queue_test.go:29]: didn't we settle on using when instead of context?

**Review by ahg-g — COMMENTED** [review]
few nits, can you please address Kante's comments so we merge this?

**Comment by ahg-g** [comment]
Hi @knight42 are you still on this one?

File: apis/kueue/v1alpha1/queue_webhook_test.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+		},
+		"status could be updated": {
+			before: testingutil.MakeQueue(testQueueName, testQueueNamespace).Obj(),
+			after:  testingutil.MakeQueue(testQueueName, testQueueNamespace).PendingWorkloads(10).Obj(),
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			errList := ValidateQueueUpdate(tc.before, tc.after)
+			if len(tc.wantErr) == 0 {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply apis/kueue/v1alpha1/queue_webhook_test.go:80]: @knight42 can you address this please as well?

[the remark this answers is not in the document]
Comment (knight42) [reply apis/kueue/v1alpha1/queue_webhook_test.go]: I have overlooked.. should be fixed this time

**Pull request merged**